godot_rl/core/godot_env.py

```diff
@@ -387,19 +387,34 @@ def _clear_socket(self):
 
     def _get_data(self):
         try:
-            data = self.connection.recv(4)
-            if not data:
-                time.sleep(0.000001)
-                return self._get_data()
-            length = int.from_bytes(data, "little")
-            string = ""
-            while len(string) != length:  # TODO: refactor as string concatenation could be slow
-                string += self.connection.recv(length).decode()
+            # Receive the size (in bytes) of the remaining data to receive
+            string_size_bytes: bytearray = bytearray()
+            received_length: int = 0
+
+            # The first 4 bytes contain the length of the remaining data
+            length: int = 4
+
+            while received_length < length:
```

Comment:
I like that we remove the recursion, but could we do something like:
```
string_size_bytes = None
while string_size_bytes is None:
    time.sleep(0.000001) # optional 
    string_size_bytes = self.connection.recv(4)
```

Reply:
I would like to ask about this part in more detail since I'm not that familiar about low level connection details and just started reading some articles while working on the implementation. Some assumptions below could be wrong or maybe they don't apply to this specific case and how Godot side deals with the connection, I'll be happy to correct the code as needed.

I was thinking about three questions specifically:
1 - When do we expect 'None' or zero bytes from `recv` to occur
2 - If the `recv` blocks, do we need the time.sleep which also blocks? 
3 - How do we make sure we receive the exact size of the data, regardless of how many bytes a single recv returns

So far I think that `None` or receiving zero bytes from `recv` usually means that the connection is ended from the remote side, and recv normally waits (and blocks) until at least some data is received, as we are currently using a blocking socket. If there's a timeout while waiting, the timeout exception should occur. Also, there is no guarantee that will receive the size of bytes specified in one `recv` call, it could be less. 

For non blocking sockets, recv would return immediately, but then the rest of the code may need alterations (and for some future update we could consider if using e.g. using async i/o or some higher level handling of the code may provide benefits, e.g. when dealing with your new parallel environments implementation).

The above assumptions were based on this article (https://docs.python.org/3/howto/sockets.html) and other research:
>Now we come to the major stumbling block of sockets - send and recv operate on the network buffers. They do not necessarily handle all the bytes you hand them (or expect from them), because their major focus is handling the network buffers. In general, they return when the associated network buffers have been filled (send) or emptied (recv). They then tell you how many bytes they handled. It is your responsibility to call them again until your message has been completely dealt with.

> When a recv returns 0 bytes, it means the other side has closed (or is in the process of closing) the connection. You will not receive any more data on this connection. Ever. You may be able to send data successfully; I’ll talk more about this later.

Regarding as to why my code is receiving the first part of the message (length) in a loop just like the actual string is received, I'm not sure if this happens here, but the consideration itself was inspired by this part of the article:

>Prefixing the message with its length (say, as 5 numeric characters) gets more complex, because (believe it or not), you may not get all 5 characters in one recv. In playing around, you’ll get away with it; but in high network loads, your code will very quickly break unless you use two recv loops - the first to determine the length, the second to get the data part of the message.

Reply:
Do you think we might be able to merge this one? While I haven't tested this extensively, the changes should probably work. I'm just wondering if we can clear out some outstanding PRs.